### PR TITLE
[interp] Add Internal.Runtime.CompilerServices.Unsafe intrinsics.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mono/mini/interp/interp.c [hunk 1]
```diff
@@ -4470,6 +4470,28 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
 			MINT_IN_BREAK;
 			break;
 		}
+		MINT_IN_CASE(MINT_INTRINS_UNSAFE_ADD_BYTE_OFFSET) {
+			sp -= 2;
+			sp [0].data.p = (guint8*)sp [0].data.p + sp [1].data.i;
+			sp ++;
+			++ip;
+			MINT_IN_BREAK;
+			break;
```
Comment (BrzVlad) [inline mono/mini/interp/interp.c]: this break (and below) is not needed

File: mono/mini/interp/interp.c [hunk 2]
```diff
@@ -4470,6 +4470,28 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
...
+		}
+		MINT_IN_CASE(MINT_INTRINS_UNSAFE_BYTE_OFFSET) {
+			sp -= 2;
+			sp [0].data.i = (guint8*)sp [1].data.p - (guint8*)sp [0].data.p;
+			sp ++;
+			++ip;
+			MINT_IN_BREAK;
+			break;
+		}
+		MINT_IN_CASE(MINT_INTRINS_UNSAFE_AS) {
```
Comment (BrzVlad) [inline mono/mini/interp/interp.c]: maybe we should emit a MINT_NOP instead. We could also ignore these later on when emitting the code.

File: mono/mini/interp/mintops.def [hunk 3]
```diff
@@ -667,3 +667,6 @@ OPDEF(MINT_PROF_ENTER, "prof_enter", 1, MintOpNoArgs)
 
 OPDEF(MINT_INTRINS_BYREFERENCE_CTOR, "intrins_byreference_ctor", 2, MintOpClassToken)
 OPDEF(MINT_INTRINS_BYREFERENCE_GET_VALUE, "intrins_byreference_get_value", 1, MintOpNoArgs)
+OPDEF(MINT_INTRINS_UNSAFE_ADD_BYTE_OFFSET, "intrins_unsafe_add_byte_offset", 2, MintOpNoArgs)
```
Comment (BrzVlad) [inline mono/mini/interp/mintops.def]: the size of these two instructions is 1

File: mono/mini/interp/transform.c [hunk 4]
```diff
@@ -1196,7 +1196,31 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoMeth
...
+		else if (!strcmp (tm, "As"))
+			*op = MINT_INTRINS_UNSAFE_AS;
+		else if (!strcmp (tm, "SizeOf")) {
+			MonoGenericContext *ctx = mono_method_get_context (target_method);
+			g_assert (ctx);
+			g_assert (ctx->method_inst);
+			g_assert (ctx->method_inst->type_argc == 1);
+			MonoType *t = ctx->method_inst->type_argv [0];
+			int align;
+			int esize = mono_type_size (t, &align);
```
Comment (lewurm) [inline mono/mini/interp/transform.c]: should we use `mini_get_underlying_type (t)` here? 😄

File: mono/mini/interp/transform.c [hunk 5]
```diff
@@ -1196,7 +1196,31 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoMeth
 			td->ip += 5;
 			return TRUE;
 		}
+	} else if (in_corlib && !strcmp (klass_name_space, "Internal.Runtime.CompilerServices") && !strcmp (klass_name, "Unsafe")) {
+#ifdef ENABLE_NETCORE
+		if (!strcmp (tm, "AddByteOffset"))
+			*op = MINT_INTRINS_UNSAFE_ADD_BYTE_OFFSET;
+		if (!strcmp (tm, "ByteOffset"))
```
Comment (lewurm) [inline mono/mini/interp/transform.c]: `else if`

Comment (jaykrell) [reply mono/mini/interp/transform.c]: It is too bad the strings don't carry their lengths with them, in an efficient to get way.

File: mono/mini/interp/transform.c (hunk 4 above)
Comment (vargaz) [reply mono/mini/interp/transform.c]: Thats only needed in code which deals with generic sharing etc.

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/interp/interp.c (hunk 2 above)
Comment (BrzVlad) [reply mono/mini/interp/interp.c]: build failed because this wasn't also removed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lewurm** [comment]
@monojenkins build Linux WebAssembly

**Comment by lewurm** [comment]
@monojenkins squash

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**